### DISCUSSION TITLE
horizontal scaling proposal

### DISCUSSION THREAD

**Discussion by luanshaotong** [labring/sealfs#136] [General]
横向扩所容分为两个场景：
## 扩容
扩容过程首先由manager向所有server发出信息（可以由server反向同步），之后server之间同步数据，并同时处理请求。
### 同步文件
由目前已有的节点控制扩容进度，每个节点扫描所有文件并记录其中将被移动到新加节点的文件。
需要一个本地锁，标记文件状态：已同步，同步中，未同步，请求中（阻塞），不存在
#### 同步完成
本地要迁移的文件全部迁移完成，会向manager发送同步完成的信号。当所有server全部发送了同步完成的信号后，就完成了整个扩容。manager通知client。
### client端
只发送请求到原目标server
### server请求处理
- 本server仍未同步到新server上线的消息 -> 正常处理所有请求
- 已同步到新server上线的消息
  - 该文件应该放到本节点 -> 正常处理请求
  - 该文件应该放到新节点
    - 文件状态为已同步 -> 转发请求到新节点
    - 文件状态为同步中 -> 转发请求到新节点
    - 文件状态为不存在 -> 转发请求到新节点
    - 文件状态为请求中或未同步
      - 读取
        - 非写入中-> 直接读取该文件
        - 写入中 -> 阻塞
      - 写入
        - 非请求中 -> 直接读取该文件
        - 请求中 -> 阻塞
      - 删除
        - 非请求中 -> 直接读取该文件
        - 请求中 -> 阻塞
## 缩容
缩容过程首先由manager向所有server发出信息（可以由server反向同步），之后server之间同步数据，并同时处理请求。
### 同步文件
由删除控制扩容进度，扫描所有文件并记录其中将被移动到其他节点的文件。
需要一个本地锁，标记文件状态：已同步，同步中，未同步，请求中（阻塞），不存在
### client server
处理逻辑同上

**Comment by uran0sH** [comment]
这个本地锁咋做？

**Reply by Tricky511** [reply]
dashmap?

**Comment by luanshaotong** [comment]
节点状态：
normal
迁移中
本server完成
所有server完成
所有client切换到新
所有server切换到新

判断：
check forward flaq
- yes
- no
  - if path in dashmap and not transfered
    - call local(old)
      - if create/delete
        - if state > client done
          - call add entry new
        - else
          - call add entry old
      - process file and meta
  - else
    - call new
      - if create/delete
        - if state > client done
          - call add entry new
        - else
          - call add entry old
      - process file and meta

**Comment by luanshaotong** [comment]
迁移的实现状态转换过程（基于cluster_status）
![image](https://github.com/labring/sealfs/assets/14962503/699a2251-f801-4f73-a125-8e77c93b8636)
1. 中括号表示client和server调用的哪个环，小括号则是对应的依赖项。由于cluster_status和环之间没有做到完全同步，是顺序执行的，所以会产生上一个阶段的请求在新阶段中使用上一个阶段的map情况，其中如果当前阶段中环的内容要变更，则需要讨论是否存在问题。
2. manager对于所有的环的更新是同步的，不存在阶段不一致的问题。
